Add Uniswap integration

bingen opened the pull request:
Allow to buy ETH or any other token.

TODO:
- [x] Add tests.

Commit:
Add Uniswap integration

Allow to buy ETH or any other token.

Commit:
Add tests

Reach 100% coverage.

izqui commented on contracts/CourtPresaleActivate.sol:
```diff
@@ -51,8 +57,72 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
+        require(_amount > 0, ERROR_ZERO_AMOUNT);
+
+        // move tokens to this contract
+        require(ERC20(_token).safeTransferFrom(_from, address(this), _amount), ERROR_TOKEN_TRANSFER_FAILED);
```
🚨 This allows anyone too see if someone creates an approval to the contract and then perform the action on their behalf

```suggestion
        require(ERC20(_token).safeTransferFrom(msg.sender, address(this), _amount), ERROR_TOKEN_TRANSFER_FAILED);
```

bingen replied:
Sorry, I don't get this one. The account doing the approval should be `_from` too, and the tokens staked and activated in the Registry will be on behalf of `from`. Am I missing something?
Precisely, using parameter `_from` here allows to perform this tx from a different account than the one that did the initial approval (I guess this could become more useful when we do the same for the bonding curve than now in the presale, but I don't see how it's harmful).

izqui commented on contracts/CourtPresaleActivate.sol:
```diff
@@ -51,8 +57,72 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
+
+        // get the Uniswap exchange for the contribution token
+        address contributionTokenAddress = address(presale.contributionToken());
+        IUniswapExchange uniswapExchange = IUniswapExchange(uniswapFactory.getExchange(contributionTokenAddress));
```
🚨 From reading the Uniswap source, it is actually the other way around. We should be calling the exchange contract for `_token` and passing `presale.contributionToken()` as the output token:

```suggestion
        IUniswapExchange uniswapExchange = IUniswapExchange(uniswapFactory.getExchange(_token));
```

izqui commented on contracts/CourtPresaleActivate.sol:
```diff
@@ -51,8 +57,72 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
+        IUniswapExchange uniswapExchange = IUniswapExchange(uniswapFactory.getExchange(contributionTokenAddress));
+
+        // swap tokens
+        uint256 contributionTokenAmount = uniswapExchange.tokenToTokenSwapInput(_amount, _minTokens, _minEth, _deadline, _token);
```
```suggestion
        uint256 contributionTokenAmount = uniswapExchange.tokenToTokenSwapInput(_amount, _minTokens, _minEth, _deadline, presale.contributionToken());
```

izqui commented on contracts/CourtPresaleActivate.sol:
```diff
@@ -51,8 +57,72 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
+        address contributionTokenAddress = address(presale.contributionToken());
+        IUniswapExchange uniswapExchange = IUniswapExchange(uniswapFactory.getExchange(contributionTokenAddress));
+
+        // swap tokens
```
🚨 We need to approve `uniswapExchange` to be able to pull `amount` from `this`

izqui commented on contracts/CourtPresaleActivate.sol:
```diff
@@ -51,8 +57,72 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
+
+        // swap tokens
+        uint256 contributionTokenAmount = uniswapExchange.ethToTokenSwapInput.value(msg.value)(_minTokens, _deadline);
+
```
⚠️ Is there any chance that we will end up with leftover ETH that we need to return to `msg.sender`?

bingen replied:
I don't think so, because we are using the [`Input` version](https://docs.uniswap.io/smart-contract-api/exchange#ethtotokenswapinput), where the ETH value is supposed to [be exact](https://github.com/Uniswap/contracts-vyper/blob/master/contracts/uniswap_exchange.vy#L145).

bingen replied:
This is done. I didn't use the "withdrawal pattern" because I don't think it's a good idea (I think that if there are leftovers it's good to revert if they cannot be sent back), but let me know if you think the opposite.

izqui commented on contracts/test/ERC20BadMock.sol:
```diff
@@ -0,0 +1,144 @@
+import "@aragon/court/contracts/standards/ApproveAndCall.sol";
+
+
+contract ERC20BadMock {
```
Can you add a comment on why is it a `BadMock`?

izqui commented on contracts/test/PresaleMock.sol:
```diff
@@ -57,6 +57,10 @@ contract PresaleMock is IPresale {
         return collateralToken;
     }
 
+    function setCollateralToken(ERC20 _collateralToken) external {
```
Why do we need this change in this PR?

bingen replied:
It's only a mock, so it was to make things easier and tests lighter, but yes, I can re-deploy everything with the new token.

izqui commented on test/buy_and_activate.js:
```diff
@@ -46,22 +57,153 @@ contract('Court presale and activate wrapper', ([_, owner, juror1, juror2, juror
+      await badExternalToken.mint(juror1, INITIAL_BIG_TOKEN_AMOUNT)
+
+      // add it to Presale
+      await presale.setCollateralToken(badCollateralToken.address)
```
Hmm, why not set up the test from scratch using `badCollateralToken` in the setup for the presale?

bingen replied:
Yes, I'll do it, see comment above.

izqui commented on contracts/test/UniswapMock.sol:
```diff
@@ -0,0 +1,121 @@
+    function tokenToEthTransferOutput(uint256 eth_bought, uint256 max_tokens, uint256 deadline, address recipient) external returns (uint256  tokens_sold) {}
+
+    // Trade ERC20 to ERC20
+    function tokenToTokenSwapInput(uint256 tokens_sold, uint256 min_tokens_bought, uint256 min_eth_bought, uint256 deadline, address token_addr) external returns (uint256  tokens_bought) {
```
🚨 This is not how Uniswap works, see Vyper implementation: https://github.com/Uniswap/contracts-vyper/blob/master/contracts/uniswap_exchange.vy

In the actual implementation, it:
- Pulls `token` from `msg.sender`
- Convert `token` to ETH in this Exchange contract
- Call `Exchange(token_addr).ethToTokenTransferInput()` which transfers `token_addr` to `msg.sender`

I recommend using an actual implementation in Solidity instead of mocking it: https://github.com/PhABC/uniswap-solidity/blob/master/contracts/uniswap/UniswapExchange.sol

Commit:
Add Uniswap: fix fetching wrong exchange

Address PR #3 comments.

Commit:
Remove _from parameter from contributeExternalToken

Address PR #3 comments.

Commit:
Add extra-safety measures

Make sure no funds are left in the contract.
Address PR #3 comments.

Commit:
fixup! Add extra-safety measures

Commit:
Add check for non-existent Uniswap exchange

Commit:
test: Fix name in aux variable

sohkai commented on contracts/lib/uniswap/uniswap/UniswapExchange.sol:
```diff
@@ -0,0 +1,613 @@
+pragma solidity ^0.5.0;
```
I'm just wondering; if we're only using these contracts in the tests (I assume so, since we already have the interface versions), can we put these in a `test/` dir?

Commit:
Move Uniswap contracts to test folder

Address PR #3 comments.

sohkai commented on contracts/CourtPresaleActivate.sol:
```diff
@@ -49,10 +59,93 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
+        }
+    }
+
+    function _buyAndActivate(address _from, uint256 _amount, address _token) internal {
```
For internal functions, we might as well enforce type checking, so we could use `ERC20 _token`.

sohkai commented on contracts/CourtPresaleActivate.sol:
```diff
@@ -62,6 +155,21 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
         emit BoughtAndActivated(_from, _token, _amount, bondedTokensObtained);
     }
+
+    function _refund(address _tokenAddress) internal {
```
Do we really need this extra jump? I think asking for a cast at the call site should be enough?

bingen replied:
With the change above, not anymore :wink:

sohkai commented on contracts/CourtPresaleActivate.sol:
```diff
@@ -49,10 +59,93 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
+        require(ERC20(_token).safeApprove(address(presale), _amount), ERROR_TOKEN_APPROVAL_FAILED);
 
         // buy in presale
         presale.contribute(address(this), _amount);
```
I don't have much context on this, but if we have the `bondedToken` in storage, is there a reason we're using `presale.contributionToTokens(_amount)` instead of `bondedToken.balanceOf(this)`?

Are they equivalent or are we protecting for the case where someone accidentally sends the bonded token to this contract (because the `presale` separately tracks the contribution amount from this contract)?

bingen replied:
Yes, they should be equivalent, unless someone accidentally sends as you said. Anyway, those would be sent on `_refund` later. Not sure what's the best option here unless we implement an escape hatch mechanism governed by AN DAO or something like that.

sohkai commented on contracts/CourtPresaleActivate.sol:
```diff
@@ -62,6 +155,21 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
         registry.stakeFor(_from, bondedTokensObtained, abi.encodePacked(ACTIVATE_DATA));
 
+        // refund leftovers if any
+        _refund(bondedToken);
```
This goes with the above question; how would someone end up with an amount of bonded tokens to need refunding? Is it only possible if someone accidentally sent the bonded token to this contract and then the next person using this contract will receive it?

bingen replied:
Yes, that, or a bug in the presale or registry contracts (that was my initial assumption, as with Uniswap).

sohkai commented on contracts/CourtPresaleActivate.sol:
```diff
@@ -49,10 +59,93 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
+
+        // refund leftovers if any
+        _refund(token);
+        _refund(contributionTokenAddress);
```
Similar to above the above questions.

At this point, we should be guaranteed to use up all of the received `token`, but the contract may have received some of `token` by mistake / random. Perhaps we should adjust the refund with the contract's starting balance of `token` before it pulls it?

And for `contributionToken`, would we ever have a situation where the presale would only accept a portion of the contribution?

-------

I'm wondering if we might want to think of a different recovery mechanism.

Given that there's no way to really track randomly deposited amounts, perhaps we can have an owned escape hatch callable by the AN DAO or etc.

bingen replied:
> Similar to above the above questions.
> 
> At this point, we should be guaranteed to use up all of the received `token`, but the contract may have received some of `token` by mistake / random. Perhaps we should adjust the refund with the contract's starting balance of `token` before it pulls it?

The initial motivation was more along the lines of bugs/inaccuracies of the external contracts, see [this comment](https://github.com/aragonone/court-presale-activate/pull/3#discussion_r367345562), but yours is a good point too.

> And for `contributionToken`, would we ever have a situation where the presale would only accept a portion of the contribution?

Theoretically not.

> I'm wondering if we might want to think of a different recovery mechanism.
> 
> Given that there's no way to really track randomly deposited amounts, perhaps we can have an owned escape hatch callable by the AN DAO or etc.

You read my mind, haha, I proposed that in a response above.

sohkai commented on contracts/CourtPresaleActivate.sol:
```diff
@@ -49,10 +59,93 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
+        // make sure there's no ETH left
+        uint256 ethBalance = address(this).balance;
+        if (ethBalance > 0) {
+            (bool result,) = msg.sender.call.value(ethBalance)("");
```
Similar to above, we may end up with ETH randomly (e.g. self destruct) and may want to adjust with the starting balance and `msg.value`.

sohkai commented on contracts/CourtPresaleActivate.sol:
```diff
@@ -49,10 +59,93 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
+    * @param _minTokens Minimum amount of presale contribution tokens obtained in Uniswap
+    * @param _deadline Transaction deadline for Uniswap
+    */
+    function contributeEth(uint256 _minTokens, uint256 _deadline) external payable {
```
Perhaps it would be useful to expose a fallback function that calls this with `now` and `min = 1` like Uniswap does?

sohkai commented on contracts/CourtPresaleActivate.sol:
```diff
@@ -49,10 +59,93 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
+        address _token,
+        uint256 _amount,
+        uint256 _minTokens,
+        uint256 _minEth,
```
Not particularly opposed to exposing `minTokens` and `minEth` like this, but I'm wondering if we should.

`minEth`, if I understand it correctly, is to control slippage across exchanges (but I don't understand why it matters, since you're controlling the end token amount anyway. Maybe for liquidity reasons?). I don't think there's much point in exposing this parameter and we could likely just use `1` where it's needed.

`minTokens` is confusing, because we have three tokens involved: external, contribution, and bonded. As a user, I'd likely be most interested in the final bonded token amount I'm getting, so it may be more useful (and less confusing) to provide that as a parameter instead.

bingen replied:
This is just forwarding the Uniswap interface:
https://docs.uniswap.io/smart-contract-api/exchange#tokentotokenswapinput
I think it makes sense to leave it as is, because in the end to understand you have to go to Uniswap docs/code, and adding our own intermediate translation would only create more confusion in my opinion.
About `minEth`, we can always make frontend to fill in a `1` in it, so I don't see any problem in having this param.

sohkai reviewed (commented):
A couple of questions / suggestions. I don't have great context on how the presale works precisely, so some of these notes may be moot.

I think we may want to re-consider how we do refunds / recoveries, mostly so we have control to help recover / be good samaritans if an accident happens.

sohkai commented on contracts/CourtPresaleActivate.sol:
```diff
@@ -49,10 +59,93 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
+    )
+        external
+    {
+        require(_amount > 0, ERROR_ZERO_AMOUNT);
```
Not actually sure if/how Uniswap handles this, but we may want to revert if the `_token` is the contribution token. I don't remember any explicit checks in the solidity version, and it'd be awful if it let you do the trade but took the fees.

If it does handle it, we might want to add a comment and / or branching later to handle this.

Commit:
Use typed ERC20 in internal functions

Commit:
Change refund mechanism

Add separate refund protected functions, allowed only to governor.
Address PR #3 comments.

Commit:
Add check for external token different than collateral

Address PR #3 comments.

Commit:
Add fallback function

Address PR #3 comments.

sohkai commented on contracts/CourtPresaleActivate.sol:
```diff
@@ -134,13 +136,34 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
+        uint256 selfBalance = address(this).balance;
+        require(selfBalance >= _amount, ERROR_NOT_ENOUGH_BALANCE);
+
+        (bool result,) = _recipient.call.value(_amount)("");
```
This is a small thing, but with solc 0.5 do you need the `("")` or can you get away with `()`?

Just seems a bit weird but I can also see making the case for using `""` to make it more explicit.

bingen replied:
Yes, without it I get:
```
Error: CompileError: /home/bingen/workspace/court_presale_activate/contracts/CourtPresaleActivate.sol:140:26: TypeError: Wrong argument count for function call: 0 arguments given but expected 1. This function requires a single bytes argument. Use "" as argument to provide empty calldata.
        (bool result,) = _recipient.call.value(_amount)();
                         ^------------------------------^
```

sohkai reviewed (approved):
LGTM! Let's merge and get the other PRs updated :).
